tensorboard/plugins/graph/tf_graph_common/render.ts

```diff
@@ -753,6 +759,44 @@ export class RenderGraphInfo {
       coreGraph.setEdge(edgeObj.v, edgeObj.w, renderMetaedgeInfo);
     });
 
+    // If there are functions, it is possible for metanodes to be dynamically
+    // added later. Construct the hierarchies for nodes that are predecessors to
+    // nodes in the current hierarchy so that edges are drawn correctly.
+    if (!_.isEmpty(this.hierarchy.libraryFunctions)) {
```

Comment:
the method `cloneFunctionLibraryMetanode` is getting pretty big. Might be a good idea to break it down into several smaller methods.

Reply:
Indeed - I will do that. https://github.com/tensorflow/tensorboard/issues/573